public/app/features/explore/state/actions.ts

```diff
@@ -401,11 +431,304 @@ export function modifyQueries(
  */
 export function runQueries(exploreId: ExploreId): ThunkResult<void> {
   return (dispatch, getState) => {
-    dispatch(updateTimeRangeAction({ exploreId }));
-    dispatch(runQueriesAction({ exploreId }));
+    dispatch(updateTime({ exploreId }));
+
+    const exploreItemState = getState().explore[exploreId];
+    const {
+      datasourceInstance,
+      queries,
+      datasourceError,
+      containerWidth,
+      isLive: live,
+      queryState,
+      queryIntervals,
+      range,
+      scanning,
+      history,
+    } = exploreItemState;
+
+    if (datasourceError) {
+      // let's not run any queries if data source is in a faulty state
+      return;
+    }
+
+    if (!hasNonEmptyQuery(queries)) {
+      dispatch(clearQueriesAction({ exploreId }));
+      dispatch(stateSave()); // Remember to save to state and update location
+      return;
+    }
+
+    // Some datasource's query builders allow per-query interval limits,
+    // but we're using the datasource interval limit for now
+    const interval = datasourceInstance.interval;
+
+    if (!queryState.isStarted()) {
+      queryState.cancel('New request issued');
+      queryState.closeStreams(false);
+    }
+    queryState.sendFrames = true;
+    queryState.sendLegacy = true; // temporary hack until we switch to PanelData
+
+    const queryOptions = { interval, maxDataPoints: containerWidth, live };
+    const datasourceId = datasourceInstance.meta.id;
+    const now = Date.now();
+    const transaction = buildQueryTransaction(queries, queryOptions, range, queryIntervals, scanning);
+
+    // temporary hack until we switch to PanelData, Loki already converts to DataFrame so using legacy will destroy the format
+    const isLokiDataSource = datasourceInstance.meta.name === 'Loki';
+
+    queryState.onStreamingDataUpdated = () => {
+      const data = queryState.validateStreamsAndGetPanelData();
+      const { state, error, legacy, series } = data;
+      if (!data && !error && !legacy && !series) {
+        return;
+      }
+
+      if (state === LoadingState.Error) {
+        dispatch(processErrorResults({ exploreId, response: error, datasourceId }));
+        return;
+      }
+
+      if (state === LoadingState.Streaming) {
+        dispatch(limitMessageRate(exploreId, isLokiDataSource ? series : legacy, datasourceId));
```

Comment:
We have to replace the call to `limitMessageRate` with  `processQueryResults` and implement throttling in datasources instead because Prometheus datasource uses this as a way to run Instant and TimeSeries queries in parallel.

Reply:
@hugohaggmark `limitMessageRate` does not actually limit anything right now does it?

Reply:
Nope because of the reason in my comment, it would break Prometheus queries.

Reply:
More details: to make Prometheus parallel querying work with PanelQueryState I use the streaming callback. But after posting each query's result I immediately send a "stop this stream" kind of result as well. If we then throttle/debounce that it will miss the first result with the data. So I'm just passing data along, I'm currently simplifying this even further in next PR so limitMessageRate will go away and also @ryantxu is adding buffering directly to the Loki datasource.

Reply:
So is there anything to do on my side? I mean seems like my work with the buffering was kinda hijacked :D. Should I just wait for @ryantxu and reuse that work afterwards?

Reply:
Sorry didn't mean to hijack your work but I think you'll have to move your part to the loki datasource for now. Let me know if you need help with that. And yes it's hard to sync 3 ongoing efforts